src/proposer/proposer.cpp

```diff
@@ -29,15 +30,83 @@
 
 namespace proposer {
 
-class ProposerStub : public Proposer {
+bool GenerateBlock(staking::ActiveChain &active_chain,
+                   staking::TransactionPicker &transaction_picker,
+                   proposer::BlockBuilder &block_builder,
+                   proposer::Logic &logic,
+                   CWallet *wallet,
+                   const CBlockIndex &tip,
+                   const staking::CoinSet &coins,
+                   const boost::optional<CScript> &coinbase_script,
+                   std::shared_ptr<const CBlock> &block_out) {
+
+  auto &wallet_ext = wallet->GetWalletExtension();
+  const auto wallet_name = wallet->GetName();
+
+  const boost::optional<EligibleCoin> &winning_ticket = logic.TryPropose(coins);
+  if (!winning_ticket) {
+    LogPrint(BCLog::PROPOSING, "Not proposing this time (wallet=%s)\n", wallet_name);
+    return false;
+  }
+  const EligibleCoin &coin = winning_ticket.get();
+  LogPrint(BCLog::PROPOSING, "Proposing... (wallet=%s, coin=%s)\n",
+           wallet_name, util::to_string(coin.utxo));
+  staking::TransactionPicker::PickTransactionsParameters parameters{};
+  staking::TransactionPicker::PickTransactionsResult result =
+      transaction_picker.PickTransactions(parameters);
+
+  if (!result) {
+    LogPrint(BCLog::PROPOSING, "Failed to pick transactions (wallet=%s, error=%s) – proposing empty block.\n", wallet_name, result.error);
+  }
+  const CAmount fees = std::accumulate(result.fees.begin(), result.fees.end(), CAmount(0));
+  const uint256 snapshot_hash = active_chain.ComputeSnapshotHash();
+
+  block_out = block_builder.BuildBlock(
+      tip, snapshot_hash, coin, coins, result.transactions, fees, coinbase_script, wallet_ext);
+
+  return block_out != nullptr;
+}
+
+class PassiveProposerImpl : public Proposer {
+ private:
+  const Dependency<staking::ActiveChain> m_active_chain;
+  const Dependency<staking::TransactionPicker> m_transaction_picker;
+  const Dependency<proposer::BlockBuilder> m_block_builder;
+  const Dependency<proposer::Logic> m_proposer_logic;
+
  public:
+  PassiveProposerImpl(Dependency<staking::ActiveChain> const active_chain,
+                      Dependency<proposer::BlockBuilder> const block_builder,
+                      Dependency<staking::TransactionPicker> const transaction_picker,
+                      Dependency<proposer::Logic> const proposer_logic)
```

Comment:
Shouldn't we accept const values instead of const pointers? E.g.`const Dependency<proposer::Logic> proposer_logic`?

Reply:
@kostyantyn `const Dependency<T>` and `Dependency<T> const` are the same — `T *const`.

Reply:
@frolosofsky thanks! learned smth new, I got confused with how the template/using is getting evaluated.

Reply:
But still, our codestyle dictates `const` to be in front, doesn't it?